.github/workflows/sync-ref.yml

```diff
@@ -0,0 +1,100 @@
+        uses: actions/github-script@v6
+        id: check
+        with:
+          script: |
+            const getSHA = async (repository, ref) => {
+              if (ref.startsWith('refs/heads/') || ref.startsWith('refs/tags/')) ref = ref.substring(4)
+              else throw new Error(`Cannot handle ref '${ref}`)
+
+              try {
+                const { data: { object: { sha } } } = await github.rest.git.getRef(
+                  repository
+                    .split('/')
+                    .reduce((map, e) => {
+                      map[map.owner ? 'repo' : 'owner'] = e
+                      return map
+                  }, { ref })
+                )
+                return sha
+              } catch (e) {
+                if (e?.status == 404) return undefined
+                throw e
+              }
+            }
+
+            const sourceSHA = await getSHA(process.env.SOURCE_REPOSITORY, process.env.REF)
+            const targetSHA = await getSHA(process.env.TARGET_REPOSITORY, process.env.REF)
+            // skip sync if SHAs match, making extra certain that `master` is also synced to `main`
+            const skip = sourceSHA !== targetSHA
+              ? false
+              : (process.env.REF !== 'refs/heads/master' ||
+                 sourceSHA === await getSHA(process.env.TARGET_REPOSITORY, 'refs/heads/main'))
+            core.setOutput('skip', skip ? 'true' : 'false')
+            core.setOutput('source-sha', sourceSHA)
+            core.setOutput('target-sha', targetSHA)
+      - name: obtain installation token
+        if: steps.check.outputs.skip == 'false'
+        uses: tibdex/github-app-token@b62528385c34dbc9f38e5f4225ac829252d1ea92
+        id: token
+        with:
+          app_id: ${{ secrets.GITGITGADGET_GITHUB_APP_ID }}
+          private_key: ${{ secrets.GITGITGADGET_GITHUB_APP_PRIVATE_KEY }}
+          repository: ${{ env.TARGET_REPOSITORY }}
+      - name: set authorization header
+        if: steps.check.outputs.skip == 'false'
+        uses: actions/github-script@v6
+        id: auth
+        with:
+          script: |
+            // Sadly, `git push` does not work with 'Authorization: Bearer <PAT>', therefore
+            // we have to use the `Basic` variant
+            const auth = Buffer.from('PAT:${{ steps.token.outputs.token }}').toString('base64')
+            core.setSecret(auth)
+            core.setOutput('header', `Authorization: Basic ${auth}`)
+      - name: sync
+        if: steps.check.outputs.skip == 'false'
+        shell: bash
+        run: |
+          set -ex
+          git init --bare
+
+          git remote add source "${{ github.server_url }}/$SOURCE_REPOSITORY"
+          # pretend to be a partial clone
+          git config remote.source.promisor true
+          git config remote.source.partialCloneFilter blob:none
+
+          # fetch some commits
+          git fetch --depth 10000 source ${{ steps.check.outputs.source-sha }}
+          rm -f .git/shallow
```

Comment:
I guess we need to be careful and wrap this:

```sh
case "${{ steps.check.outputs.source-sha }}" in
''|undefined) ;; # the ref was deleted; nothing needs to be fetched
*)
  [...]
  ;;
esac
```

because it is possible that `git/git` removes branches that then need to be removed from gitgitgadget/git, too.

Reply:
Any reason you did not push this?  I assume full support for branch deletion would be a separate workflow.

Reply:
Heh, I forgot to fix it locally ;-) Will do so in a minute.